File: apps/repl/src/tools/treaty-sfc/treaty/parser.ts
import { Token, ControlFlowKind, DeferKind, TokenType } from './token';

interface AstNode {
    type: TokenType;
    value: string;
}

interface Ast {
    nodes: AstNode[];
}

class Parser {
    private tokens: Token[];
    private pos: number;

    constructor(tokens: Token[]) {
        this.tokens = tokens;
        this.pos = 0;
    }

    parse(): Ast {
        const nodes: AstNode[] = [];

        while (!this.isAtEnd()) {
            const node = this.parseNode();
            nodes.push(node);
        }

        return { nodes };
    }

    private parseNode(): AstNode {
        const token = this.advance();

        switch (token.kind.type) {
            case TokenType.JavaScript:
                return { type: TokenType.JavaScript, value: token.kind.value };
            case TokenType.Style:
                return { type: TokenType.Style, value: token.kind.value };
            case TokenType.HTML:
                return { type: TokenType.HTML, value: token.kind.value };
            case TokenType.TemplateExpression:
                return { type: TokenType.HTML, value: token.kind.value };
            case TokenType.ControlFlow:
                return this.parseControlFlowNode(token.kind.kind);
            case TokenType.Defer:
                return this.parseDeferNode(token.kind.kind);
            case 'Eof':
                return { type: 'EOF', value: '' };
            default:
                throw new Error(`Unexpected token type: ${token.kind}`);
        }
    }

    private parseControlFlowNode(kind: ControlFlowKind): AstNode {
        switch (kind) {
            case ControlFlowKind.If:
                return { type: TokenType.HTML, value: '@if' };
            case ControlFlowKind.ElseIf:
                return { type: TokenType.HTML, value: '@else if' };
            case ControlFlowKind.Else:
                return { type: TokenType.HTML, value: '@else' };
            case ControlFlowKind.For:
                return { type: TokenType.HTML, value: '@for' };
            case ControlFlowKind.Empty:
                return { type: TokenType.HTML, value: '@empty' };
            case ControlFlowKind.Switch:
                return { type: TokenType.HTML, value: '@switch' };
            case ControlFlowKind.Case:
                return { type: TokenType.HTML, value: '@case' };
            case ControlFlowKind.Default:
                return { type: TokenType.HTML, value: '@default' };
        }
    }

    private parseDeferNode(kind: DeferKind): AstNode {
        switch (kind) {
            case DeferKind.Defer:
                return { type: TokenType.HTML, value: '@defer' };
            case DeferKind.Placeholder:
                return { type: TokenType.HTML, value: '@placeholder' };
            case DeferKind.Loading:
                return { type: TokenType.HTML, value: '@loading' };
            case DeferKind.Error:
                return { type: TokenType.HTML, value: '@error' };
        }
    }
    private advance(): Token {
        if (!this.isAtEnd()) {
            this.pos++;
        }
        return this.previous();
    }

    private isAtEnd(): boolean {
        return this.peek().kind.type === 'Eof';
    }

    private peek(): Token {
        return this.tokens[this.pos] || { kind: { type: 'Eof' }, start: 0, end: 0 };
    }

    private previous(): Token {
        return this.tokens[this.pos - 1];
    }
}

export { Parser, type AstNode, type Ast };

File: apps/rust/authoring/src/treaty/parser.rs
use crate::treaty::token::{ControlFlowClause, Token, TokenKind};
use crate::treaty::ast::{Ast, AstNode, ControlFlowBranch};

pub struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Parser { tokens, pos: 0 }
    }

    /// Parses the tokens and returns an AST.
    pub fn parse(&mut self) -> Ast {
        let mut nodes = Vec::new();

        while !self.is_at_end() {
            let node = self.parse_node();
            nodes.push(node);
        }

        Ast { nodes }
    }

    fn parse_node(&mut self) -> AstNode {
        let token = self.advance();
        let token_kind = token.kind.clone(); // Clone the token kind to avoid multiple mutable borrows

        match token_kind {
            TokenKind::JavaScript(code) => AstNode::JavaScript(code),
            TokenKind::Style { content, lang } => AstNode::Style { content, lang },
            TokenKind::Macro { content, info } => AstNode::Macro { content, info },
            TokenKind::HTML(content) => AstNode::Html(content),
            TokenKind::TemplateExpression(expr) => AstNode::TemplateExpression(expr),
            TokenKind::ControlFlow { verbatim, clauses } => {
                AstNode::ControlFlow { verbatim, clauses: Self::parse_clauses(clauses) }
            }
            TokenKind::ServerBlock { lang, body, verbatim } => {
                AstNode::ServerBlock { lang, body, verbatim }
            }
            TokenKind::Eof => AstNode::EOF,
        }
    }

    /// Turn the lexer's structured control-flow clauses into AST branches, parsing each clause's
    /// recursively-lexed body token stream into child AST nodes — so a control-flow region is a REAL
    /// nested AST (its body markup / JS / NESTED control flow become children), not a flat marker.
    fn parse_clauses(clauses: Vec<ControlFlowClause>) -> Vec<ControlFlowBranch> {
        clauses
            .into_iter()
            .map(|clause| {
                let ControlFlowClause { keyword, kind, head, body } = clause;
                // Recurse: parse the clause body's token stream into child AST nodes via a sub-parser.
                // The lexer never emits an explicit `Eof` token (the body token stream ends naturally),
                // and `Parser::parse` terminates on the synthetic out-of-bounds `Eof` from `peek`, so
                // the body tokens are parsed as-is with no trailing sentinel node.
                let body_nodes = Parser::new(body).parse().nodes;
                ControlFlowBranch { keyword, kind, head, body: body_nodes }
            })
            .collect()
    }

    fn advance(&mut self) -> &Token {
        if !self.is_at_end() {
            self.pos += 1;
        }
        self.previous()
    }

    fn is_at_end(&self) -> bool {
        self.peek().kind == TokenKind::Eof
    }

    fn peek(&self) -> &Token {
        self.tokens.get(self.pos).unwrap_or(&Token {
            kind: TokenKind::Eof,
            start: 0,
            end: 0,
        })
    }

    fn previous(&self) -> &Token {
        self.tokens.get(self.pos - 1).unwrap()
    }
}